src/lib/mixer_module/mixer_module.cpp

```diff
@@ -0,0 +1,525 @@
+		if (_groups_subscribed & (1 << i)) {
+			_control_subs[i].set_interval_us(_max_topic_update_interval_us);
+		}
+	}
+}
+
+void MixingOutput::setAllMinValues(uint16_t value)
+{
+	for (unsigned i = 0; i < MAX_ACTUATORS; i++) {
+		_min_value[i] = value;
+	}
+}
+
+void MixingOutput::setAllMaxValues(uint16_t value)
+{
+	for (unsigned i = 0; i < MAX_ACTUATORS; i++) {
+		_max_value[i] = value;
+	}
+}
+
+void MixingOutput::setAllFailsafeValues(uint16_t value)
+{
+	for (unsigned i = 0; i < MAX_ACTUATORS; i++) {
+		_failsafe_value[i] = value;
+	}
+}
+
+void MixingOutput::setAllDisarmedValues(uint16_t value)
+{
+	for (unsigned i = 0; i < MAX_ACTUATORS; i++) {
+		_disarmed_value[i] = value;
+	}
+}
+
+void MixingOutput::unregister()
+{
+	for (auto &control_sub : _control_subs) {
+		control_sub.unregister_callback();
+	}
+}
+
+bool MixingOutput::update()
+{
+	if (!_mixers) {
+		handleCommands();
+		// do nothing until we have a valid mixer
+		return false;
+	}
+
+	// check arming state
+	if (_armed_sub.update(&_armed)) {
+		_armed.in_esc_calibration_mode &= _support_esc_calibration;
+		/* Update the armed status and check that we're not locked down.
+		 * We also need to arm throttle for the ESC calibration. */
+		_throttle_armed = (_safety_off && _armed.armed && !_armed.lockdown) || (_safety_off && _armed.in_esc_calibration_mode);
+	}
+
+	if (_param_mot_slew_max.get() > FLT_EPSILON) {
+		const hrt_abstime now = hrt_absolute_time();
+		const float dt = math::constrain((now - _time_last_mix) / 1e6f, 0.0001f, 0.02f);
+		_time_last_mix = now;
+
+		// maximum value the outputs of the multirotor mixer are allowed to change in this cycle
+		// factor 2 is needed because actuator outputs are in the range [-1,1]
+		const float delta_out_max = 2.0f * 1000.0f * dt / (_max_value[0] - _min_value[0]) / _param_mot_slew_max.get();
+		_mixers->set_max_delta_out_once(delta_out_max);
+	}
+
+	unsigned n_updates = 0;
+
+	/* get controls for required topics */
+	for (unsigned i = 0; i < actuator_controls_s::NUM_ACTUATOR_CONTROL_GROUPS; i++) {
+		if (_groups_subscribed & (1 << i)) {
+			if (_control_subs[i].copy(&_controls[i])) {
+				n_updates++;
+			}
+
+			/* During ESC calibration, we overwrite the throttle value. */
+			if (i == 0 && _armed.in_esc_calibration_mode) {
+
+				/* Set all controls to 0 */
+				memset(&_controls[i], 0, sizeof(_controls[i]));
+
+				/* except thrust to maximum. */
+				_controls[i].control[actuator_controls_s::INDEX_THROTTLE] = 1.0f;
+
+				/* Switch off the PWM limit ramp for the calibration. */
+				_pwm_limit.state = PWM_LIMIT_STATE_ON;
+			}
+		}
+	}
+
+	/* do mixing */
+	float outputs[MAX_ACTUATORS] {};
+	const unsigned mixed_num_outputs = _mixers->mix(outputs, MAX_ACTUATORS);
+
+	/* the PWM limit call takes care of out of band errors, NaN and constrains */
+	uint16_t output_limited[MAX_ACTUATORS] {};
+
+	pwm_limit_calc(_throttle_armed, armNoThrottle(), mixed_num_outputs, _reverse_output_mask,
```

Comment:
Is there anything PWM specific left here? pwm_limit_calc -> limit_calc

Reply:
I don't see anything. The diff will be quite large though, but I can do it.

Reply:
Done.